src/embind/emval.js

```diff
@@ -383,12 +383,18 @@ var LibraryEmVal = {
     return id;
   },
 
+  $emval_registeredMethods: [],
   _emval_get_method_caller__sig: 'iii',
-  _emval_get_method_caller__deps: ['_emval_addMethodCaller', '_emval_lookupTypes', '$new_', '$makeLegalFunctionName'],
+  _emval_get_method_caller__deps: ['_emval_addMethodCaller', '_emval_lookupTypes', '$new_', '$makeLegalFunctionName', '$emval_registeredMethods'],
   _emval_get_method_caller: function(argCount, argTypes) {
     var types = __emval_lookupTypes(argCount, argTypes);
-
     var retType = types[0];
+    var signatureName = retType.name + "_$" + types.slice(1).map(function (t) { return t.name; }).join("_") + "$";
```

Comment:
I'm not sure how reliable type names are (can they clash with each other?). Perhaps safer to use typeids instead.

Reply:
Do we have access to typeids here? Also, don't we expect 1-1 mapping between rawType and the type name (since I didn't entirely follow how the type name was derived via template metaprogramming on the C++ side)

If type names aren't unique, then I am thinking of the following but there is likely a simpler way
```
function __emval_getRawTypes(idx, argTypes) { // new method
      return requireRegisteredType(HEAP32[(argTypes >> 2) + idx]);
}

var signatureName = // get raw types for args and return type and concatenate
```
does this seem reasonable?

Reply:
IIRC (but I haven't looked at it in a while), the names are given programmatically rather than derived, so it's technically possible for them to clash. You can recheck though if you want.

Personally, I'd go for typeids as they're guaranteed to be unique. The ids are in the `argTypes` array as you noticed, so you could just do something like

```js
HEAPU32.subarray(argTypes >> 2, (argTypes >> 2) + argCount).toString()
```

which would give you a string with comma-separated typeids, which is a perfectly unique key.

Reply:
Ahh I only just noticed that you didn't introduce this code but only moved it around from a different place. I guess if it worked before, then it's fine...

Reply:
Doesn't hurt to be more correct I suppose, I can make this change regardless; wdyt?

Reply:
I would actually avoid changing that in this PR.   Better to keep changes separate.  Maybe as a followup.. but also maybe its find to leave it as is?